File: neetcode/6. linked list/11.designBrowserHistory.py
class ListNode:
  def __init__(self, val, prev=None, next=None):
    self.val = val
    self.prev = prev
    self.next = next

class BrowserHistory:
  def __init__(self, homepage: str):
    self.cur = ListNode(homepage)
  
  def visit(self, url: str):
    self.cur.next = ListNode(url, self.cur)
    self.cur = self.cur.next

  def back(self, steps: int) -> str:
    while self.cur.prev and steps > 0:
      self.cur = self.cur.prev
      steps -= 1
    return self.cur.val
  
  def forward(self, steps: int) -> str:
    while self.cur.next and steps > 0:
      self.cur = self.cur.next
      steps -= 1
    return self.cur.val

###############################################
# solution using an array/stack
class BrowserHistory:
  def __init__(self, homepage: str):
    self.i = 0  # i is the current pointer
    self.len = 1
    self.history = [homepage]
  
  def visit(self, url: str) -> None:
    if len(self.history) < self.i + 2:
      self.history.append(url)
    else:
      self.history[self.i+1] = url
    self.i += 1
    self.len = self.i + 1
  
  def back(self, steps: int) -> str:
    self.i = max(self.i - steps, 0)
    return self.history[self.i]
  
  def forward(self, steps: int) -> str:
    self.i = min(self.i + steps, self.len - 1)
    return self.history[self.i]

File: neetcode/6. linked list/11.designBrowserHistory.cpp
// You have a browser of one tab where you start on the homepage and you can visit another url, 
// get back in the history number of steps or move forward in the history number of steps.

// Implement the BrowserHistory class:

// BrowserHistory(string homepage) Initializes the object with the homepage of the browser.
// void visit(string url) Visits url from the current page. It clears up all the forward history.
// string back(int steps) Move steps back in history. If you can only return x steps in the history 
// and steps > x, you will return only x steps. Return the current url after moving back in history 
// at most steps.
// string forward(int steps) Move steps forward in history. If you can only forward x steps in the 
// history and steps > x, you will forward only x steps. Return the current url after forwarding 
// in history at most steps.

#include <string>
#include <stack>
using namespace std;

// solution using 2 stacks
class BrowserHistory {
public:
  stack<string> history;
  stack<string> future;
  
  BrowserHistory(string homepage) {
    history.push(homepage);
    future = stack<string>(); //resets stack
  }
  
  void visit(string url) {
    history.push(url);
    future = stack<string>();
  }
  
  string back(int steps) {
    while(steps>0 && history.size() > 1){
      future.push(history.top());
      history.pop();
      steps--;
    }
    return history.top();
  }
  
  string forward(int steps) {
    while(steps>0 && future.size()>0){
      history.push(future.top());
      future.pop();
      steps--;
    }
    return history.top();
  }
};

// solution using doubly linked list (more efficient)
class BrowserHistory {
public:
  class Node{
    public:
    string url;
    Node* next;
    Node* prev;
    
    Node(string url){
      this->url =url;
      next = NULL;
      prev = NULL;
    }
  };
  Node* curr;
  BrowserHistory(string homepage) {
    curr = new Node(homepage);
  }
  
  void visit(string url) {
    Node* temp = new Node(url);
    curr->next = temp;
    temp->prev = curr;
    curr = curr->next;
  }
  
  string back(int steps) {
    while(curr->prev!=nullptr && steps-- >0){
      curr = curr->prev;
    }
    return curr->url;
  }
  
  string forward(int steps) {
    while(curr->next!=nullptr && steps-- >0){
      curr = curr->next;
    }
    return curr->url;
  }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */